include/aikido/control/Executor.hpp

```diff
@@ -0,0 +1,130 @@
+#ifndef AIKIDO_CONTROL_EXECUTOR_HPP_
+#define AIKIDO_CONTROL_EXECUTOR_HPP_
+
+#include <chrono>
+#include <future>
+#include <set>
+#include <vector>
+
+#include "aikido/common/ExecutorThread.hpp"
+#include "aikido/common/pointers.hpp"
+
+namespace aikido {
+namespace control {
+
+AIKIDO_DECLARE_POINTERS(Executor)
+
+enum class ExecutorType
+{
+  STATE = -1,
+  POSITION = 0,
+  VELOCITY = 1,
+  EFFORT = 2,
+  TRAJECTORY = 3,
+  OTHER = 4
+};
+
+// Parameter defaults
+constexpr std::chrono::milliseconds defaultThreadRate{10};
+
+/// Abstract class for executing commands
+class Executor
+{
+public:
+  virtual ~Executor()
+  {
+    stop();
+  }
+
+  /// Constructor.
+  /// \param[in] types Vector of controller types
+  /// \param[in] joints Vector of joint names this Executor acts upon
+  /// \param[in] threadRate (Optional) How often to call step()
+  Executor(
+      const std::vector<ExecutorType> types,
+      const std::vector<std::string> joints,
+      std::chrono::milliseconds threadRate = defaultThreadRate)
+    : mThreadRate(threadRate), mThread(nullptr), mTypes(types), mJoints(joints)
+  {
+  }
+
+  Executor(
+      const ExecutorType type,
+      const std::vector<std::string> joints,
+      std::chrono::milliseconds threadRate = defaultThreadRate)
+    : Executor(std::vector<ExecutorType>{type}, joints, threadRate)
+  {
+  }
+
+  /// Get primary ExecutorType
+  ExecutorType getType()
+  {
+    return mTypes[0];
+  }
+
+  /// Get all of this Executor's ExecutorTypes
+  std::vector<ExecutorType> getTypes()
```

Comment:
Main use case: consider a `RosRobot` that handles controller switching. When you give it an executor, it will check the types and joint names to see which controller it needs to spin up.